modules/tensor_mechanics/src/materials/crystal_plasticity/CrystalPlasticityHCPDislocationSlipBeyerleinUpdate.C

```diff
@@ -0,0 +1,602 @@
+    _forest_dislocations_removed_increment(
+        declareProperty<std::vector<Real>>(_base_name + "_forest_dislocations_removed_increment")),
+    _initial_forest_dislocation_density(getParam<Real>("initial_forest_dislocation_density")),
+    _total_substructure_density(declareProperty<Real>(_base_name + "total_substructure_density")),
+    _total_substructure_density_old(
+        getMaterialPropertyOld<Real>(_base_name + "total_substructure_density")),
+    _total_substructure_density_increment(
+        declareProperty<Real>(_base_name + "total_substructure_increment")),
+    _initial_substructure_density(getParam<Real>("initial_substructure_density")),
+
+    _slip_system_modes(getParam<unsigned int>("slip_system_modes")),
+    _number_slip_systems_per_mode(
+        getParam<std::vector<unsigned int>>("number_slip_systems_per_mode")),
+    _lattice_friction(getParam<std::vector<Real>>("lattice_friction_per_mode")),
+
+    _reference_strain_rate(getParam<Real>("gamma_o")),
+    _rate_sensitivity_exponent(getParam<Real>("strain_rate_sensitivity_exponent")),
+
+    _burgers_vector(getParam<std::vector<Real>>("burgers_vector_per_mode")),
+    _slip_generation_coefficient(
+        getParam<std::vector<Real>>("slip_generation_coefficient_per_mode")),
+    _slip_activation_energy(
+        getParam<std::vector<Real>>("normalized_slip_activiation_energy_per_mode")),
+    _proportionality_factor(
+        getParam<std::vector<Real>>("slip_energy_proportionality_factor_per_mode")),
+    _forest_interaction_coefficient(getParam<Real>("forest_interaction_parameter")),
+    _boltzman_constant(getParam<Real>("Boltzman_constant")),
+    _macro_applied_strain_rate(getParam<Real>("applied_strain_rate")),
+    _macro_reference_strain_rate(getParam<Real>("reference_macroscopic_strain_rate")),
+
+    _shear_modulus(getParam<std::vector<Real>>("effective_shear_modulus_per_mode")),
+    _substructure_rate_coefficient(
+        getParam<std::vector<Real>>("substructure_rate_coefficient_per_mode")),
+    _substructure_hardening_coefficient(getParam<Real>("substructure_hardening_coefficient")),
+    _hallpetch_like_coefficient(getParam<std::vector<Real>>("Hall_Petch_like_constant_per_mode")),
+    _grain_size(getParam<Real>("grain_size"))
+{
+  // resize local caching vectors used for substepping
+  _previous_substep_slip_resistance.resize(_number_slip_systems);
+  _previous_substep_forest_dislocations.resize(_number_slip_systems);
+  _slip_resistance_before_update.resize(_number_slip_systems);
+  _forest_dislocations_before_update.resize(_number_slip_systems);
+}
+
+void
+CrystalPlasticityHCPDislocationSlipBeyerleinUpdate::initQpStatefulProperties()
+{
+  // check that the number of slip systems is equal to the sum of the types of slip system
+  if (_number_slip_systems_per_mode.size() != _slip_system_modes)
+    mooseError("The size the number of slip systems per mode is not equal to the number of slip "
```

Comment:
Maybe using `paramError` is better here (and in the following)?

Reply:
done, thanks!